src/pages/en/guides/client-side-scripts.md

```diff
@@ -0,0 +1,205 @@
+
+## Common script patterns
+
+### Handle `onclick` and other events
+
+Some UI frameworks use custom syntax for event handling like `onclick={...}` (React/Preact) or `@click="..."` (Vue). Astro components follow web standards and recommend using [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) in a `<script>` tag to handle user interactions.
+
+```astro title="src/components/AlertButton.astro"
+<button data-alert>Click me!</button>
+
+<script>
+  // Find all buttons with the `data-alert` attribute on the page.
+  const buttons = document.querySelectorAll('button[data-alert]');
+
+  // Handle clicks on each button.
+  buttons.forEach((button) => {
+    button.addEventListener('click', () => {
+      alert('Button was clicked!');
+    });
+  });
+</script>
+```
+
+:::note
+Because this script will run once even if we add many instances of `<AlertButton />` to a page, we use `querySelectorAll` to get all instances of our component and add listeners to each one instead of using `querySelector`, which would only return the first button it found.
+:::
+
+### Web components with custom elements
+
+[Custom elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) provide a standard API that allows you to define your own HTML elements with custom behavior. Defining a custom element in a `.astro` component can be a good way to build interactivity without reaching for a UI framework library.
+
+In this example, we define a new `<astro-heart>` HTML element that tracks how many times you click the heart button and updates the `<span>` with the latest count.
+
+```astro title="src/components/AstroHeart.astro"
+<!-- Wrap the component elements in our custom element “astro-heart”. -->
+<astro-heart>
+  <button aria-label="Heart">💜</button> × <span>0</span>
+</astro-heart>
+
+<script>
+  // Define the behaviour for our new type of HTML element.
+  class AstroHeart extends HTMLElement {
+    constructor() {
+			super();
+      let count = 0;
+
+      const heartButton = this.querySelector('button');
+      const countSpan = this.querySelector('span');
+
+      // Each time the button is clicked, update the count.
+			heartButton.addEventListener('click', () => {
+        count++;
+        countSpan.textContent = count;
+      });
+		}
+  }
+
+  // Tell the browser to use our AstroHeart class for <astro-heart> elements.
+  customElements.define('astro-heart', AstroHeart);
+</script>
+```
+
+This is similar to how you might handle user input without a custom element. But, instead of using `document.querySelector()`, we used `this.querySelector()` to get DOM elements. `this.querySelector()` only searches within the current custom element instance, so it’s easier to work with the children of one component instance at a time. Because the browser understands that this code is for a custom element, it will automatically run it for you each time it finds `<astro-heart>` on the page.
```

Comment:
Flagging as needs work!  😅 

I'm not sure *I* get why this is different than the previous case, other than "your browser understands that this code is for a custom element"... so in that spirit, I'm going to try something (that you will correct later) like...
```suggestion
 Unlike scripts that are bundled and only run once, the script for `<astro-heart>` will be run by the browser each time it finds this custom element on the page. The browser understands that `this.querySelector()` only searches within the current custom element instance to get DOM elements. So, even if you have multiple `<AstroHeart />` components, its script does not have to query for/add event listenters to all instances  of its child elements.
```

Reply:
Yeah — I wasn’t happy with this either. Here’s a second attempt.

```suggestion
There are two advantages to using a custom element here:

1. You can use `this.querySelector()` to get DOM elements instead of searching the whole page using `document.querySelector()`. `this.querySelector()` only searches within the current custom element instance, making it easier to work with the children of one component instance at a time.

2. The browser will run our custom element’s `constructor()` method each time it finds `<astro-heart>` on the page. This means your code only needs to handle one component at a time instead of setting up all components on the page at once.
```

Will commit this so I can see it better in context and await feedback.